sinks/factory.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sinks
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"strconv"
+
+	"github.com/GoogleCloudPlatform/heapster/extpoints"
+	sink_api "github.com/GoogleCloudPlatform/heapster/sinks/api"
+	"github.com/GoogleCloudPlatform/heapster/sinks/gcl"
+	"github.com/GoogleCloudPlatform/heapster/sinks/gcm"
+	"github.com/GoogleCloudPlatform/heapster/sinks/influxdb"
+	"github.com/golang/glog"
+)
+
+func init() {
+	extpoints.SinkFactories.Register(CreateInfluxdbSink, "influxdb")
+	extpoints.SinkFactories.Register(CreateGCMSink, "gcm")
+	extpoints.SinkFactories.Register(CreateGCLSink, "gcl")
+}
+
+func CreateInfluxdbSink(uri string, options map[string][]string) ([]sink_api.ExternalSink, error) {
+	defaultConfig := influxdb.Config{
+		User:         "root",
+		Password:     "root",
+		Host:         "localhost:8086",
+		DbName:       "k8s",
+		AvoidColumns: false,
+	}
+
+	parsedUrl, err := url.Parse(os.ExpandEnv(uri))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(parsedUrl.Host) > 0 {
+		defaultConfig.Host = parsedUrl.Host
+	}
+	if len(options["user"]) >= 1 {
+		defaultConfig.User = options["user"][0]
+	}
+	if len(options["pw"]) >= 1 {
+		defaultConfig.Password = options["pw"][0]
+	}
+	if len(options["db"]) >= 1 {
+		defaultConfig.DbName = options["db"][0]
+	}
+	if len(options["avoidColumns"]) >= 1 {
+		val, err := strconv.ParseBool(options["avoidColumns"][0])
+		if err != nil {
+			return nil, fmt.Errorf("invalid value %q for option 'avoidColumns' passed to influxdb sink", options["avoidColumns"][0])
+		}
+		defaultConfig.AvoidColumns = val
+	}
+	sink, err := influxdb.New(defaultConfig)
+	if err != nil {
+		return nil, err
+	}
+	glog.Infof("creating influxdb sink with options: %v", defaultConfig)
+
+	return []sink_api.ExternalSink{sink}, nil
+}
+
+func CreateGCMSink(_ string, _ map[string][]string) ([]sink_api.ExternalSink, error) {
+	sink, err := gcm.New()
```

Comment:
By flags are you referring to the opaque part of the url? These drivers will require more options once they are enhanced to work outside of GCE as well.